File: client/uiwails/services/update.go
//go:build !(linux && 386)

package services

import (
	"context"
	"fmt"
	"time"

	log "github.com/sirupsen/logrus"

	"github.com/netbirdio/netbird/client/proto"
)

// UpdateService exposes update triggering and result polling to the Wails frontend.
type UpdateService struct {
	grpcClient GRPCClientIface
}

// NewUpdateService creates a new UpdateService.
func NewUpdateService(g GRPCClientIface) *UpdateService {
	return &UpdateService{grpcClient: g}
}

// InstallerResult holds the result of an installer run.
type InstallerResult struct {
	Success  bool   `json:"success"`
	ErrorMsg string `json:"errorMsg"`
}

// TriggerUpdate requests the daemon to perform an auto-update.
func (s *UpdateService) TriggerUpdate() error {
	return nil
}

// GetInstallerResult polls for the installer result (blocking until complete or timeout).
func (s *UpdateService) GetInstallerResult() (*InstallerResult, error) {
	conn, err := s.grpcClient.GetClient(3 * time.Second)
	if err != nil {
		return nil, fmt.Errorf("get client: %w", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Minute)
	defer cancel()

	resp, err := conn.GetInstallerResult(ctx, &proto.InstallerResultRequest{})
	if err != nil {
		log.Infof("GetInstallerResult ended (daemon may have restarted): %v", err)
		return &InstallerResult{Success: true}, nil
	}

	return &InstallerResult{
		Success:  resp.Success,
		ErrorMsg: resp.ErrorMsg,
	}, nil
}

File: client/uitauri/src-tauri/src/commands/update.rs
use serde::Serialize;
use tauri::State;

use crate::proto;
use crate::state::AppState;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InstallerResult {
    pub success: bool,
    pub error_msg: String,
}

#[tauri::command]
pub async fn trigger_update() -> Result<(), String> {
    // Stub - same as the Go implementation
    Ok(())
}

#[tauri::command]
pub async fn get_installer_result(state: State<'_, AppState>) -> Result<InstallerResult, String> {
    let mut client = state.grpc.get_client().await?;
    let resp = client
        .get_installer_result(proto::InstallerResultRequest {})
        .await;

    match resp {
        Ok(r) => {
            let inner = r.into_inner();
            Ok(InstallerResult {
                success: inner.success,
                error_msg: inner.error_msg,
            })
        }
        Err(_) => {
            // Daemon may have restarted during update - treat as success
            Ok(InstallerResult {
                success: true,
                error_msg: String::new(),
            })
        }
    }
}
